Fix off-by-one warning from LGTM and add tests for NodePort range

christarazi opened the pull request:
<!-- Description of change -->

Fixes warning as seen here: https://lgtm.com/projects/g/cilium/cilium/snapshot/5770266093fca7b394682a3a2ba6a21f92ec07da/files/pkg/option/config.go?sort=name&dir=ASC&mode=heatmap#xd94e5a86fe539522:1

https://github.com/cilium/cilium/blob/dd534bc55947d54b8071ae067238ae9cdbe1b66b/pkg/option/config.go#L1654

Side note: I was looking to add test coverage relevant to the change and noticed that [DaemonConfig.Populate()](https://github.com/cilium/cilium/blob/dd534bc55947d54b8071ae067238ae9cdbe1b66b/pkg/option/config.go#L1654) is not tested. After investigating a bit, I came to the conclusion that I'll cover it in a separate PR. This is because `Populate()` does not return an error and instead does `log.Fatal` when something goes wrong. This made writing unit tests to cover non-happy path more difficult. If refactoring `Populate()` to return an `error` seems reasonable, I will go ahead with that.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10151)
<!-- Reviewable:end -->

aanm reviewed (approved):
Thanks for the PR @christarazi !

christarazi committed:
Refactor NodePortRange code into separate function

This change separates out the NodePortRange logic from Populate() into a
separate function to enable easier unit testing. This commit also brings
in said unit tests.

Signed-off-by: Chris Tarazi <tarazichris@gmail.com>

christarazi committed:
Validate NodePortRange length

We can rely on the defaults to set NodePortRange to a slice of length 2
when it has NOT been passed in by the user explicitly. If NodePortRange
is passed in explicitly, then we ensure that it must be of length 2. If
it is explicitly empty, we simply log a warning.

Signed-off-by: Chris Tarazi <tarazichris@gmail.com>

joestringer commented on pkg/option/config.go:
```diff
@@ -1964,7 +1964,12 @@ func (c *DaemonConfig) populateNodePortRange() error {
-		return errors.New("Unable to parse min/max port value for NodePort range!")
+		// TODO(ctz): This check allows NodePortRange to be passed as explicity
+		// empty, which the pkg/envoy_test.go expects. Should that be the case
+		// or does pkg/envoy_test.go need to be modified?
```
I'm a bit surprised that this is the case. /cc @jrajahalme do you know why `pkg/envoy_test.go` would rely on this?

christarazi replied:
After taking another look, I'm going to take a stab at explaining. 

envoy_test.go uses a "preconfigured" `option.Config`: https://github.com/cilium/cilium/blob/640b7a6cb35905ca1aafd3010baf9b8265d383c9/pkg/envoy/envoy_test.go#L68

Then it calls `Populate` which will set additional values in `Config` with values from viper. Due to envoy_test.go running in the test context, the defaults *for* viper haven't been set. They haven't been set because the code which does set defaults is in `daemon_main.go`'s `init()`:

https://github.com/cilium/cilium/blob/5b9002d61500b674c04b5bdd4aa00f91619af8e0/daemon/daemon_main.go#L187

In a non-testing context (running normally), the `init()` mentioned above would run before `Populate` gets called, and everything is OK.

Therefore, in the testing context `Populate` runs standalone meaning that the `init()` mentioned above does NOT get executed. This results in three things: there are no defaults, it has nothing to retrieve from viper, and NodePortRange stays empty.

This is why you can see in the unit tests I wrote (d8f9160), all the pre-test setups are setting up defaults to inform viper, simply because the input to the function under test is viper itself.  

The only reason envoy_test.go complained is because the unit tests I wrote checked for an empty NodePortRange. The code before this PR silently allowed that, whether by accident or on purpose. Either way, it is quite tricky to wrap your head around, but I hope that cleared it up. 

It feels wrong to allow NodePortRange to be empty. I think this all just may have been a coincidental result. We can modify envoy_test.go to reduce its dependency on the entire `DaemonConfig` struct, and only just the options it needs. However I recognize that we may be getting into yak shaving territory here. Let me know what you think.

joestringer replied:
Oh, yeah this seems a bit broken for `pkg/envoy` unit-testing to be relying upon CLI args, but I'm not all that into yak-shaving so it's probably fine to leave it as-is.

There's no big danger here that the guard rails arounnd CLI validation cause inexperienced users to waste lots of time chasing down issues with this, the nodeport range is a pretty advanced configuration and anyone configuring it should likely have an idea about the correct way to pass the values in.

joestringer commented on pkg/option/config.go:
```diff
@@ -1959,6 +1947,34 @@ func (c *DaemonConfig) Populate() {
+		// TODO(ctz): This check allows NodePortRange to be passed as explicity
+		// empty, which the pkg/envoy_test.go expects. Should that be the case
+		// or does pkg/envoy_test.go need to be modified?
+		if len(nodePortRange) != 0 {
```
nit: this `if` condition can be folded into the `else` check above.

christarazi replied:
Fixed: cd702be

joestringer reviewed (changes requested):
CI is happy now :tada: 

The added unit tests make this much easier to look at and validate, thanks for adding those. The separate refactoring commit also made review easier.

I think you can drop the first two commits, and with that the commit messages in the later commits don't really make sense given that the code doesn't use `viper.IsSet()`.  I found the explanation useful specifically for reviewing this PR, but it's probably not necessary to leave in the codebase just given how infrequently we need to deal with this detail of the library.

I don't have an answer for you on the expectations from the envoy_test, but given your explanation that the defaults we configure will set the array's values correctly then it's probably valid for the final patch's functional changes to be omitted? Feel free to ping me on Slack tomorrow if you want a tighter turn-around with the CI.

christarazi commented on pkg/option/config.go:
```diff
@@ -1959,6 +1947,32 @@ func (c *DaemonConfig) Populate() {
+		// TODO(ctz): This check allows NodePortRange to be passed as explicity
+		// empty, which the pkg/envoy_test.go expects. Should that be the case
+		// or does pkg/envoy_test.go need to be modified?
+		return errors.New("Unable to parse min/max port value for NodePort range!")
```
This error string is directly used as the log message in the caller. I can change this if people would like.

joestringer reviewed (approved):
Looks like it hit known CI flake #9297 , this is not the same as the failure that we were seeing earlier in this PR development. LGTM, will just wait until we open the merge window for v1.8 development PRs.

aanm reviewed (commented):
@christarazi it seems we are introducing the usage of `errors.Errorf` with this change. Is there a reason for this? cc @joestringer
